exporter/loadbalancingexporter/loadbalancer.go

```diff
@@ -96,6 +96,7 @@ func newLoadBalancer(params component.ExporterCreateSettings, cfg config.Exporte
 		res:              res,
 		componentFactory: factory,
 		exporters:        map[string]component.Exporter{},
+		ring:             newHashRing([]string{}),
```

Comment:
It seems it prevents the crash. I spent a bit of time to understand the overall code flow in load balancer initialization and it's not fully clear to me how it works. I am a little bit worried that this ring might be not updated in some cases. Also, first couple of batches might be not load balanced as anticipated. 

I am wondering if maybe we should synchronously initialize the resolvers. I.e. instead of defining the empty ring, maybe we should just call `lb.onBackendChanges()` in `Start` (other than just adding the callback). Or we might want to handle that in resolvers start() synchronously (blocking). This will lead to other consequences (e.g. when resolvers are stuck it might block the start of the collector). 

Lets try to ping the original author. @BinaryHexer - do you have suggestions how this should be handled?

Reply:
Sorry I was away for some time so couldn't check the notifications. I'll take a look at this PR shortly!

Reply:
This change looks okay, here is why:

- The ring is being initialized during loadbalancer creation before calling `Start` method.
- The `Start` method just registers a callback from the resolver that updates the resolved endpoints.
- `onBackendChanges` is invoked by the resolver, which will create and compare the new and existing hash ring.

>  I am a little bit worried that this ring might be not updated in some cases. Also, first couple of batches might be not load balanced as anticipated.

In case of a stuck resolver or any another case that doesn't trigger `onBackendChanges` there will be no endpoints registers and hence no exporters. Effectively `Endpoint()` will return `""` as the result which when fed as input to the `Exporter()` method will trigger  an error.

The ring update is controlled by the resolver which will invoke `onBackendChanges` as explained above.

Reply:
On further checking, this change unfortunately is still not safe because of how `newHashRing()` works.  `newHashRing()` when fed an empty string slice will return a `hashRing` which still has a `nil` items field.

As we can see below, feeding an empty slice of endpoints to the below function returns a nil `ringItem`.
```go
// positionsForEndpoints calculates all the positions for all the given endpoints
func positionsForEndpoints(endpoints []string, weight int) []ringItem {
	var items []ringItem
	positions := map[position]bool{} // tracking the used positions
	for _, endpoint := range endpoints {
		// for this initial implementation, we don't allow endpoints to have custom weights
		for _, pos := range positionsFor(endpoint, weight) {
			// if this position is occupied already, skip this item
			if _, found := positions[pos]; found {
				continue
			}
			positions[pos] = true

			item := ringItem{
				pos:      pos,
				endpoint: endpoint,
			}
			items = append(items, item)
		}
	}

	sort.Slice(items, func(i, j int) bool {
		return items[i].pos < items[j].pos
	})

	return items
}```

Reply:
Disregard the above comment, there is a guard check in `findEndpoint` to ensure if the ring is empty, then it returns an empty string.

That said the title of the PR is misleading because as highlighted above the `h.items` is still `nil` after this fix.

Reply:
Awesome, many thanks for checking this @BinaryHexer

Reply:
> That said the title of the PR is misleading

Yea the real problem became more evident to me after discussions here. I didn't realize I could edit a title to a PR! Will do so now.

Reply:
After going through this and checking the PR itself, I believe this isn't the right solution. The right solution would be to do a null check at hashRing#endpointFor, like this:

```go
	if h == nil {
		// perhaps the ring itself couldn't get initialized yet?
		return ""
	}
```

This would have a similar effect, but won't synthetically initialize the ring. This line proposed on this PR would make it ambiguous to determine whether the ring hasn't been initialized or whether the list of nodes is empty.

Reply:
I believe I tried/considered this, but there's a few other places in the code that simply assume the hashRing exists and is initialized. Rather than fixing all those places to be nil-aware (and handle the resulting errors), it seemed easier (safer?) to simply make sure the hashring is properly initialized (even if it's empty).

Reply:
It might be easier, but it wouldn't be the correct place.